tools/deleteIncompleteSubmissions.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+/**
+ * @file tools/deleteIncompleteSubmissions.php
+ *
+ * Copyright (c) 2014-2022 Simon Fraser University
+ * Copyright (c) 2003-2022 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class deleteIncompleteSubmissions
+ * @ingroup tools
+ *
+ * @brief CLI tool to delete incomplete submissions
+ */
+
+use APP\facades\Repo;
+
+require(dirname(__FILE__) . '/bootstrap.inc.php');
+
+class IncompleteSubmissionDeletionTool extends CommandLineTool
+{
+    /**
+     * Constructor.
+     *
+     * @param array $argv command-line arguments
+     */
+    public function __construct($argv = [])
+    {
+        parent::__construct($argv);
+
+        // If no parameters, show usage
+        if (!sizeof($this->argv)) {
+            $this->usage();
+            exit(1);
+        }
+
+        $this->days = $argv[1];
+        $this->dryrun = false;
+
+        if (count($argv) > 2) {
+            if ($argv[2] === '--dryrun') {
+                $this->dryrun = $argv[2];
+            } else {
+                echo "Unknown argument\n";
+                exit;
+            }
+        }
+    }
+
+    /**
+     * Print command usage information.
+     */
+    public function usage()
+    {
+        echo "Permanently removes incomplete submissions.\n"
+            . "Usage: {$this->scriptName} days [--dryrun]\n"
+            . "\t\tdays: The number of days since the submission was last active, for example 365 is all submissions older than 1 year\n"
+            . "\t\t--dryrun: Only list the incomplete submission id's to be removed\n";
+    }
+
+    /**
+     * Delete submission data and associated files
+     */
+    public function execute()
+    {
+        if (!is_numeric($this->days)) {
+            echo "Number of days has to be numeric\n";
+            exit(1);
+        }
+
+
+        // Fetch all incomplete submission that are older than $this->days days
+        $contextDao = \APP\core\Application::getContextDAO();
+        $contexts = $contextDao->getAll();
+        while ($context = $contexts->next()) {
+            $collector = Repo::submission()->getCollector()->filterByContextIds([$context->getData('id')])
+                ->limit(50)
```

Comment:
was there some performance issues if we do not have a limit? no? Not sure if it was me or Dulip who wrote that part.

Reply:
I think paging ever-growing lists is always positive (even though this probably isn't going to grow that fast), I'd just move from 50 to 1000 records :)

It would be nice to batch-delete the records, but as it's a command-line tool, I don't think it's important. So at maximum a rate limiter/sleep (to not make the database too busy with the commands) would be interesting.

Just to confirm, does the code delete the related submission files? I've skimmed the OJS code fast and didn't see it.

If for some reason the delete fails (e.g. delete blocked by a foreign key), this loop could end in an infinite loop, but I guess an `Exception` will be thrown here and break everything :)

Reply:
I think this would be fine without the `->limit(50)` a version of this tool compatible with OJS `3.3` has recently been merged and is being used in production